homeassistant/components/application_credentials/__init__.py

```diff
@@ -252,7 +252,7 @@ async def _async_config_entry_app_credentials(
             item[CONF_DOMAIN] == config_entry.domain
             and item.get(CONF_AUTH_DOMAIN, item_id) == auth_domain
         ):
-            return item_id
+            return item_id  # type: ignore[no-any-return]
```

Comment:
Can't we set the generic type of the collection so that we know what type the item is?

Reply:
Only in theory at the moment, unfortunately.

First we'd need to type `storage_collection` with something like this
```py
storage_collection: ApplicationCredentialsStorageCollection = hass.data[DOMAIN][DATA_STORAGE]
```
_An example where #103844 would be quite helpful IMO._

That in turn inherits from `DictStorageCollection` which is only partially typed - missing the generic typing - currently. Even then, it could only really be typed with a TypedDict, as the value types might not always be the same, which is incompatible with `dict`. It's still something to work out tbh.

--
I would be fine to defer / close this PR if you'd like the collection typing to be fully figured out first. Not sure when that will happen though.

Reply:
I think it's better to start with the central parts that need work first. I'd look at typing `DictStorageCollection` with a base TypedDict that implementors can inherit and extend as the generic type.

Reply:
A first idea
- #108389

Reply:
This likely can be resolved without an Intersection type. Going to close this PR.